File: backend/app/services/inventory_service.py
"""
Inventory Service
Calculates on-hand inventory and manages items
"""

from sqlalchemy.orm import Session
from datetime import datetime
import uuid
from typing import Dict, List

from app.models import ConsumptionEvent, InventoryItem


class InventoryService:
    """Inventory calculations and management"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def calculate_on_hand(
        self,
        location_id: uuid.UUID,
        as_of: datetime = None
    ) -> List[Dict]:
        """Calculate on-hand inventory for all items"""
        
        if as_of is None:
            as_of = datetime.utcnow()
        
        items = self.db.query(InventoryItem).filter(
            InventoryItem.location_id == location_id,
            InventoryItem.active == True
        ).all()
        
        results = []
        
        for item in items:
            # Sum all events for this item
            events = self.db.query(ConsumptionEvent).filter(
                ConsumptionEvent.inventory_item_id == item.id,
                ConsumptionEvent.event_ts <= as_of
            ).all()
            
            on_hand = sum(e.quantity_delta for e in events)
            
            # Get current price
            current_price = item.get_current_price(as_of)
            
            results.append({
                "inventory_item_id": item.id,
                "item_name": item.name,
                "quantity": on_hand,
                "uom": item.base_uom.value,
                "unit_cost": current_price,
                "total_value": on_hand * current_price if current_price else None,
                "as_of_date": as_of
            })
        
        return results

File: packages/api/src/services/inventory.service.ts
/**
 * Inventory Service
 * Calculates on-hand inventory from the immutable ledger
 *
 * Ported from: backend/app/services/inventory_service.py
 */

import { Prisma } from "@prisma/client";
import type { ExtendedPrismaClient } from "@barstock/database";

interface OnHandItem {
  inventoryItemId: string;
  itemName: string;
  quantity: number;
  uom: string;
  unitCost: number | null;
  totalValue: number | null;
  asOfDate: Date;
}

export class InventoryService {
  constructor(private prisma: ExtendedPrismaClient) {}

  /**
   * Calculate on-hand inventory for all active items at a location
   */
  async calculateOnHand(
    locationId: string,
    asOf: Date = new Date()
  ): Promise<OnHandItem[]> {
    const items = await this.prisma.inventoryItem.findMany({
      where: { locationId, active: true },
      include: {
        priceHistory: {
          where: {
            effectiveFromTs: { lte: asOf },
            OR: [
              { effectiveToTs: null },
              { effectiveToTs: { gt: asOf } },
            ],
          },
          orderBy: { effectiveFromTs: "desc" },
          take: 1,
        },
      },
    });

    const results: OnHandItem[] = [];

    for (const item of items) {
      // Sum all consumption events for this item up to asOf
      const aggregate = await this.prisma.consumptionEvent.aggregate({
        where: {
          inventoryItemId: item.id,
          eventTs: { lte: asOf },
        },
        _sum: { quantityDelta: true },
      });

      const onHand = Number(aggregate._sum.quantityDelta ?? 0);
      const currentPrice = item.priceHistory[0]
        ? Number(item.priceHistory[0].unitCost)
        : null;

      results.push({
        inventoryItemId: item.id,
        itemName: item.name,
        quantity: onHand,
        uom: item.baseUom,
        unitCost: currentPrice,
        totalValue:
          currentPrice !== null ? onHand * currentPrice : null,
        asOfDate: asOf,
      });
    }

    return results;
  }
}
